File: code/binding/development/tools/cast/src/c++/cast/architecture/LocalWorkingMemoryAttachedComponent.cpp
/*
 * CAST - The CoSy Architecture Schema Toolkit
 *
 * Copyright (C) 2006-2007 Nick Hawes
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

#include "LocalWorkingMemoryAttachedComponent.hpp"


using namespace std;
using namespace boost;

namespace cast {

  using namespace cdl;

  LocalWorkingMemoryAttachedComponent::LocalWorkingMemoryAttachedComponent(const string &_id) 
    : //InspectableComponent(_id),
    CASTProcessingComponent(_id),
    m_pExistsConnector(NULL),
    m_pVersionNumberConnector(NULL)
  {    
  }

  void 
  LocalWorkingMemoryAttachedComponent::setPullConnector(const string & _connectionID, 
							PullConnectorOut< int > * _pOut) {
    m_pVersionNumberConnector = _pOut;
  }


  void 
  LocalWorkingMemoryAttachedComponent::setPullConnector(const string & _connectionID, 
					       PullConnectorOut< bool > * _pOut) {
    m_pExistsConnector = _pOut;
  }

  bool 
  LocalWorkingMemoryAttachedComponent::existsOnWorkingMemory(const string & _id) {

    assert(!_id.empty());//id must not be empty
    assert(m_pExistsConnector != NULL);
    FrameworkQuery query(getProcessIdentifier(),_id);
    FrameworkLocalData< bool > * pData = NULL;
    m_pExistsConnector->pull(query,pData);
    bool exists = pData->getData();
    delete pData;
    return exists;

  }

  int 
  LocalWorkingMemoryAttachedComponent::getVersionNumber(const string & _id) throw (DoesNotExistOnWMException){

    assert(m_pVersionNumberConnector != NULL);    
    assert(!_id.empty());//id must not be empty

    FrameworkQuery query(getProcessIdentifier(),_id);
    FrameworkLocalData< int > * pData = NULL;
    
    m_pVersionNumberConnector->pull(query,pData);
    int versionNumber = pData->getData();
    delete pData;

    if (-1 == versionNumber) {
      throw(DoesNotExistOnWMException(makeAdress(subarchitectureID(), _id),
				      __HERE__, 
				      "Entry has never existed on wm. Was looking in subarch %s for id %s", 
				      m_subarchitectureID.c_str(),_id.c_str()));
    }
    else {
      return versionNumber;
    }

  }  

  void 
  LocalWorkingMemoryAttachedComponent::storeVersionNumber(const string & _id, int _version) {
    m_versionNumbers[_id] = _version;
  }

  int 
  LocalWorkingMemoryAttachedComponent::getStoredVersionNumber(const string & _id) const
    throw(ConsistencyException) {

    IntMap::const_iterator i = m_versionNumbers.find(_id);

    if (i == m_versionNumbers.end()) {
      throw(ConsistencyException(makeAdress(subarchitectureID(),_id),
				 __HERE__,
				 "No stored version for id: %s", 
				 _id.c_str()));
    } else {
      return i->second;
    }
  }

  void
  LocalWorkingMemoryAttachedComponent::increaseStoredVersion(const string & _id) throw(ConsistencyException) {
    int stored = getStoredVersionNumber(_id);
    updateVersion(_id, ++stored);
  }

  void 
  LocalWorkingMemoryAttachedComponent::startVersioning(const string & _id) {
    assert (!isVersioned(_id));
    storeVersionNumber(_id, 0);
  }

  void 
  LocalWorkingMemoryAttachedComponent::updateVersion(const string & _id, int _newVersion) {
    storeVersionNumber(_id, _newVersion);
  }

  
  bool
  LocalWorkingMemoryAttachedComponent::isVersioned(const string & _id) const {
    IntMap::const_iterator i = m_versionNumbers.find(_id);
    return i != m_versionNumbers.end();
  }

  void 
  LocalWorkingMemoryAttachedComponent::stopVersioning(const string & _id) {
    if (isVersioned(_id)) {
      removeVersionNumber(_id);
    }
  }

  void 
  LocalWorkingMemoryAttachedComponent::removeVersionNumber(const string & _id) {
    m_versionNumbers.erase(_id);
  }

     
  bool 
  LocalWorkingMemoryAttachedComponent::haveLatestVersion(const string &_id)
    throw(ConsistencyException, DoesNotExistOnWMException) {
    assert (isVersioned(_id));
    
    int ownedVersion = getStoredVersionNumber(_id);
    int wmVersion = getVersionNumber(_id);
    
    debug("haveLatestVersion(%s): %d == %d", _id.c_str(), wmVersion, ownedVersion);

    return wmVersion == ownedVersion;
  }


  void 
  LocalWorkingMemoryAttachedComponent::checkConsistency(const string & _id) 
    throw(ConsistencyException, DoesNotExistOnWMException)
  {

    if (!isVersioned(_id)) {
      throw(ConsistencyException(makeAdress(subarchitectureID(),_id),
				 __HERE__,
				 "!isVersioned(%s) in subarch %s. ",
				 _id.c_str(), m_subarchitectureID.c_str()));
    }


    if (!haveLatestVersion(_id)) {
      throw(ConsistencyException(makeAdress(subarchitectureID(),_id),
				 __HERE__,
				 "%s has attempted to overwrite an outdated working memory entry. Please reread and try again. WMA: %s:%s. Local version: %d. WM version: %d",
				 getProcessIdentifier().c_str(), _id.c_str(), m_subarchitectureID.c_str(), getStoredVersionNumber(_id), getVersionNumber(_id)));
    }

  }



  void 
  LocalWorkingMemoryAttachedComponent::setPullConnector(const string & _connectionID, 			
							PullConnectorOut< WorkingMemoryPermissions > * _pOut) {
    m_lockConnector = _pOut;
  }


  void LocalWorkingMemoryAttachedComponent::lockEntry(const std::string & _id, const cdl::WorkingMemoryPermissions & _permissions)
    throw(DoesNotExistOnWMException) {
    lockEntryHelper(_id, getSubarchitectureID(), _permissions, BLOCKING);
  }

  bool LocalWorkingMemoryAttachedComponent::lockEntryHelper(const std::string & _id, 
							    const std::string & _subarch,
							    const cdl::WorkingMemoryPermissions & _permissions, 
							    const cdl::OperationMode & _op) 
    throw(DoesNotExistOnWMException) {


    assert(!_id.empty());//id must not be empty
    assert(m_lockConnector!= NULL);
    FrameworkQuery query(getProcessIdentifier(), lockQueryString(_id, _subarch, _permissions, _op));
    FrameworkLocalData< WorkingMemoryPermissions > * pData = NULL;
    m_lockConnector->pull(query,pData);
    WorkingMemoryPermissions set = pData->getData();
    delete pData;

    if (set == DOES_NOT_EXIST) {
      throw(DoesNotExistOnWMException(makeAdress(_subarch,_id),
				      __HERE__,
				      "ID %s does not exist on WM in subarch %s",_id.c_str(),_subarch.c_str()));
    }
    
    bool succeeded = (_permissions == set);
    
    // if we blocked then we should always succeed
    if (_op == BLOCKING) {
      assert(succeeded);
    }
    
    // if we succeeded, then let's store the permissions
    if (succeeded) {
      m_permissions->setPermissions(_id, _subarch, set);
    }
    
    return succeeded;
    
  }

    /**
   * Overrides the configure method from FrameworkProcess to use
   * _config to set the subarchitecture ID.
   * 
   * @param _config
   *            The ID of the subarchitecture which contains this
   *            component.
   */
  void LocalWorkingMemoryAttachedComponent::configure(std::map<std::string,std::string> & _config) {
    CASTProcessingComponent::configure(_config);
    m_permissions = 
      boost::shared_ptr< CASTComponentPermissionsMap >(new CASTComponentPermissionsMap(getSubarchitectureID()));
  }


  bool LocalWorkingMemoryAttachedComponent::tryLockEntry(const std::string & _id,
							 const cdl::WorkingMemoryPermissions & _permissions) 
    throw(DoesNotExistOnWMException) {
    return lockEntryHelper(_id, getSubarchitectureID(), _permissions,
			   cdl::NON_BLOCKING);
  }

  

  void LocalWorkingMemoryAttachedComponent::unlockEntry(const std::string & _id) 
    throw(DoesNotExistOnWMException) {
    unlockEntryHelper(_id, getSubarchitectureID());
  }
  
  void LocalWorkingMemoryAttachedComponent::unlockEntryHelper(const std::string & _id, 
							      const std::string & _subarch) 
    throw(DoesNotExistOnWMException) {
      
    // if
    if (!holdsLock(_id, _subarch)) {
      debug("no lock held for: %s:$s", _id.c_str(),_subarch.c_str());
      return;
    }

    assert(!_id.empty());//id must not be empty
    assert(m_lockConnector!= NULL);
    FrameworkQuery query(getProcessIdentifier(), unlockQueryString(_id, _subarch));
    FrameworkLocalData< WorkingMemoryPermissions > * pData = NULL;
    m_lockConnector->pull(query,pData);
    WorkingMemoryPermissions set = pData->getData();
    delete pData;

    if (set == DOES_NOT_EXIST) {
      throw(DoesNotExistOnWMException(makeAdress(_subarch,_id),
				      __HERE__,
				      "ID %s does not exist on WM in subarch %s",_id.c_str(),_subarch.c_str()));
    }
    // allow for DOES_NOT_EXIST as this could happen if unlock on
    // deletion occurs
    // boolean succeeded = (WorkingMemoryPermissions.UNLOCKED == set ||
    // WorkingMemoryPermissions.DOES_NOT_EXIST == set);
    bool succeeded = (UNLOCKED == set);
    
    // we should always succeed unlocking if we hold a lock
    assert (succeeded);
    
    m_permissions->removePermissions(_id, _subarch);

    //just check up on ourselves
     assert (!holdsLock(_id, _subarch));
	
  }

  bool LocalWorkingMemoryAttachedComponent::holdsLock(const std::string & _id) const {
    return holdsOverwriteLock(_id);
  }

  bool LocalWorkingMemoryAttachedComponent::holdsLock(const std::string & _id, const std::string & _subarch) const {
    return holdsOverwriteLock(_id, _subarch);
  }

  bool LocalWorkingMemoryAttachedComponent::holdsOverwriteLock(const std::string & _id) const {
    // any lock is an overwrite lock
    return m_permissions->hasPermissions(_id);
  }

  bool LocalWorkingMemoryAttachedComponent::holdsDeleteLock(const std::string & _id) const {
      
    if(m_permissions->hasPermissions(_id)) {
      WorkingMemoryPermissions permissions = m_permissions ->getPermissions(_id);
      return (permissions == LOCKED_OD 
	      || permissions == LOCKED_ODR);
    }
    else {
      return false;
    }
  }

  bool LocalWorkingMemoryAttachedComponent::holdsReadLock(const std::string & _id) const {
    if(m_permissions->hasPermissions(_id)) {
      WorkingMemoryPermissions permissions = m_permissions ->getPermissions(_id);
      return permissions == LOCKED_ODR;
    }
    else {
      return false;
    }
  }

  bool LocalWorkingMemoryAttachedComponent::holdsOverwriteLock(const std::string & _id, const std::string & _subarch) const {
    // any lock is an overwrite lock
    return m_permissions->hasPermissions(_id, _subarch);
  }

  bool LocalWorkingMemoryAttachedComponent::holdsDeleteLock(const std::string & _id, const std::string & _subarch) const {
    if(m_permissions->hasPermissions(_id, _subarch)) { 
      WorkingMemoryPermissions permissions = m_permissions->getPermissions(_id, _subarch);
      return (permissions == LOCKED_OD || permissions == LOCKED_ODR);
    }
      
    else {
      return false;
    }      
  }

  bool LocalWorkingMemoryAttachedComponent::holdsReadLock(const std::string & _id, const std::string & _subarch) const {
    if(m_permissions->hasPermissions(_id, _subarch)) { 
      WorkingMemoryPermissions permissions = m_permissions->getPermissions(_id, _subarch);
      return permissions == LOCKED_ODR;
    }
    else {
      return false;
    }

  }
  
  bool LocalWorkingMemoryAttachedComponent::needsConsistencyCheck(const std::string & _id) const {
    return m_permissions->needsConsistencyCheck(_id);
  }
  
  bool LocalWorkingMemoryAttachedComponent::needsConsistencyCheck(const std::string & _id, const std::string & _subarch) const {
    return m_permissions->needsConsistencyCheck(_id, _subarch);
  }
  
  void LocalWorkingMemoryAttachedComponent::consistencyChecked(const std::string & _id) {
    m_permissions->consistencyChecked(_id);
  }
  
  void LocalWorkingMemoryAttachedComponent::consistencyChecked(const std::string & _id, const std::string & _subarch) {
    m_permissions->consistencyChecked(_id, _subarch);
  }


  cdl::WorkingMemoryPermissions 
  LocalWorkingMemoryAttachedComponent::getPermissionsHelper(const std::string & _id,
							    const std::string & _subarch) const
    throw(DoesNotExistOnWMException) {

    assert(!_id.empty());//id must not be empty
    assert(m_lockConnector!= NULL);
    FrameworkQuery query(getProcessIdentifier(), statusQueryString(_id, _subarch));
    FrameworkLocalData< WorkingMemoryPermissions > * pData = NULL;
    m_lockConnector->pull(query,pData);
    WorkingMemoryPermissions permissions = pData->getData();
    delete pData;

    if (permissions == DOES_NOT_EXIST) {
      throw(DoesNotExistOnWMException(makeAdress(_subarch,_id),
				      __HERE__,
				      "ID %s does not exist on WM in subarch %s",_id.c_str(), _subarch.c_str()));
    }
    
    return permissions;
  }
  
  cdl::WorkingMemoryPermissions 
  LocalWorkingMemoryAttachedComponent::getPermissions(const std::string & _id) const
    throw(DoesNotExistOnWMException) {
    return getPermissionsHelper(_id, getSubarchitectureID());
  }


  bool 
  LocalWorkingMemoryAttachedComponent::isOverwritable(const std::string & _id) const 
    throw(DoesNotExistOnWMException) {
    if(holdsOverwriteLock(_id)) {
      return true;
    }
    else {
      const WorkingMemoryPermissions & permissions = getPermissions(_id);
      return overwriteAllowed(permissions);
    }
  }
    
  bool 
  LocalWorkingMemoryAttachedComponent::isDeletable(const std::string & _id) const 
    throw(DoesNotExistOnWMException) {
    if(holdsDeleteLock(_id)) {
      return true;
    }
    else {
      const WorkingMemoryPermissions & permissions = getPermissions(_id);
      return deleteAllowed(permissions);
    }
  }

  bool 
  LocalWorkingMemoryAttachedComponent::isReadable(const std::string & _id) const
    throw(DoesNotExistOnWMException) {
    if(holdsReadLock(_id)) {
      return true;
    }
    else {
      const WorkingMemoryPermissions & permissions = getPermissions(_id);
      return readAllowed(permissions);
    }
  }

  
}

File: binding/development/tools/cast/src/java/cast/architecture/abstr/LocalWorkingMemoryAttachedComponent.java
/**
 * 
 */
package cast.architecture.abstr;

import java.util.Hashtable;
import java.util.Properties;

import balt.core.connectors.FrameworkConnectionException;
import balt.core.connectors.pull.primitive.interfaces.BoolPullInterface.BoolPullConnectorOut;
import balt.core.connectors.pull.primitive.interfaces.BoolPullInterface.BoolPullSender;
import balt.core.connectors.pull.primitive.interfaces.IntPullInterface.IntPullConnectorOut;
import balt.core.data.FrameworkQuery;
import cast.architecture.subarchitecture.ConsistencyException;
import cast.architecture.subarchitecture.DoesNotExistOnWMException;
import cast.architecture.subarchitecture.SubarchitectureProcessException;
import cast.architecture.subarchitecture.WMException;
import cast.cdl.OperationMode;
import cast.cdl.WorkingMemoryAddress;
import cast.cdl.WorkingMemoryPermissions;
import cast.core.CASTUtils;
import cast.core.components.CASTProcessingComponent;
import cast.core.data.CASTComponentPermissionsMap;
import cast.core.interfaces.PermissionsPullInterface.PermissionsPullConnectorOut;
import cast.core.interfaces.PermissionsPullInterface.PermissionsPullSender;

/**
 * The absolute simplest component that can be attached to a working memory.
 * This component is able to check whether an entry exists on a local working
 * memory or not, and check and record the version number of the entry. This
 * class also provides a facility for storing these version numbers.
 * 
 * @author nah
 */
public abstract class LocalWorkingMemoryAttachedComponent
		extends
			CASTProcessingComponent
		implements
			BoolPullSender,
			PermissionsPullSender {

	private BoolPullConnectorOut m_existsConnector;

	private PermissionsPullConnectorOut m_lockConnector;

	protected IntPullConnectorOut m_versionNumberConnector;

	private final Hashtable<String, Integer> m_versionNumbers;

	/**
	 * A map holding the version numbers of entries used in the past.
	 * 
	 * TODO how to garbage collect this?
	 */
	private final Hashtable<String, Integer> m_oldVersionNumbers;

	/**
	 * @param _id
	 */
	public LocalWorkingMemoryAttachedComponent(String _id) {
		super(_id);
		m_existsConnector = null;
		m_lockConnector = null;
		m_versionNumbers = new Hashtable<String, Integer>();
		m_oldVersionNumbers = new Hashtable<String, Integer>();
	}

	@Override
	public void configure(Properties _config) {
		super.configure(_config);
		m_permissions = new CASTComponentPermissionsMap(getSubarchitectureID());
	}

	/**
	 * Determines whether an entry exists on working memory at the given local
	 * address.
	 * 
	 * @param _id
	 *            The id for the entry in working memory.
	 * @return True if the entry exists, otherwise false.
	 */
	public final boolean existsOnWorkingMemory(String _id)
			throws SubarchitectureProcessException {

		assert _id.length() > 0 : "id must not be empty";

		assert (m_existsConnector != null);
		try {
			return m_existsConnector.pull(new FrameworkQuery(
					getProcessIdentifier(), _id));
		}
		catch (FrameworkConnectionException e) {
			throw new SubarchitectureProcessException(
					"Unable to check existance", e);
		}
	}

	public final void setPullConnector(String _connectionID,
			BoolPullConnectorOut _senderAdaptor) {
		m_existsConnector = _senderAdaptor;
	}

	/**
	 * Get a count of the number of times the working memory entry at the given
	 * address has been overwritten.
	 * 
	 * @param _id
	 *            The id for the entry in working memory.
	 * @return The overwrite count of the entry. This will be 0 if it has never
	 *         been overwritten.
	 * @throws DoesNotExistOnWMException
	 *             if the entry has never existed on working memory.
	 * @throws SubarchitectureProcessException
	 *             if communication goes awry
	 * 
	 * 
	 * @remark Interface change: Renamed to reflect new role, behaviour still
	 *         the same: getOverwriteCount -> getVersionNumber.
	 */
	public final int getVersionNumber(String _id)
			throws DoesNotExistOnWMException, SubarchitectureProcessException {
		assert (m_versionNumberConnector != null);
		assert _id.length() > 0 : "id must not be empty";
		int version = -1;
		try {
			version = m_versionNumberConnector.pull(new FrameworkQuery(
					getProcessIdentifier(), _id));
		}
		catch (FrameworkConnectionException e) {
			throw new SubarchitectureProcessException(
					"Unable to get overwrite count", e);
		}

		if (version < 0) {
			throw new DoesNotExistOnWMException(new WorkingMemoryAddress(_id,
					getSubarchitectureID()),
					"id has never existed for version check: " + _id);
		}
		else {
			return version;
		}
	}

	public final void setPullConnector(String _connectionID,
			IntPullConnectorOut _senderAdaptor) {
		m_versionNumberConnector = _senderAdaptor;
	}

	/**
	 * Associate the given id with the given version number. This should not be
	 * called by user code.
	 * 
	 * @param _id
	 * @param _version
	 */
	protected final void storeVersionNumber(String _id, int _version) {
		m_versionNumbers.put(_id, _version);
	}

	/**
	 * Get the version number currently stored in the component for the wm id.
	 * 
	 * @param _id
	 * @return
	 * @throws ConsistencyException
	 *             if the id is not stored.
	 */
	protected final int getStoredVersionNumber(String _id)
			throws ConsistencyException {

		Integer v = m_versionNumbers.get(_id);
		if (v == null) {
			throw new ConsistencyException(new WorkingMemoryAddress(_id,
					getSubarchitectureID()), "No stored version for id: " + _id);
		}
		else {
			return v;
		}
	}
	/**
	 * Add the given id to the versioning system with version 0.
	 * 
	 * @param _id
	 */
	protected final void startVersioning(String _id) {
		assert (!isVersioned(_id));
		storeVersionNumber(_id, 0);
	}

	/**
	 * Updates the stored id with a new version number.
	 * 
	 * @param _id
	 * @param _newVersion
	 */
	protected final void updateVersion(String _id, int _newVersion) {
		if (m_bDebugOutput) {
			try {
				debug("updateVersion: " + _id + " "
						+ getStoredVersionNumber(_id) + " -> " + _newVersion);
			}
			catch (ConsistencyException e) {
				debug(e.getLocalizedMessage());
			}
		}
		storeVersionNumber(_id, _newVersion);
	}

	/**
	 * Increment the stored version number by 1.
	 * 
	 * @param _id
	 *            id of entry
	 * @throws ConsistencyException
	 *             if _id is not stored in the versioning system
	 */
	protected final void increaseStoredVersion(String _id)
			throws ConsistencyException {
		int stored = getStoredVersionNumber(_id);
		updateVersion(_id, ++stored);
	}

	/**
	 * Check whether the given id is versioned.
	 * 
	 * @param _id
	 * @return
	 */
	protected final boolean isVersioned(String _id) {
		return m_versionNumbers.containsKey(_id);
	}

	/**
	 * Removes the given id from the versioning system.
	 * 
	 * @param _id
	 */
	protected final void stopVersioning(String _id) {
		if (isVersioned(_id)) {
			removeVersionNumber(_id);
		}
	}

	/**
	 * Removes the given id from the internal versioning system storage.
	 * 
	 * @param _id
	 */
	protected final void removeVersionNumber(String _id) {
		m_versionNumbers.remove(_id);
	}

	/**
	 * Checks whether this component has read the more recent version of the
	 * data at this working memory address.
	 * 
	 * @param _id
	 * @return
	 * @throws ConsistencyException
	 *             if the id is not versioned.
	 * @throws DoesNotExistOnWMException
	 *             if the _id does not exist on wm
	 * @throws SubarchitectureProcessException
	 *             if communication goes awry
	 */
	protected final boolean haveLatestVersion(String _id)
			throws ConsistencyException, DoesNotExistOnWMException,
			SubarchitectureProcessException {
		assert (isVersioned(_id));

		int ownedVersion = getStoredVersionNumber(_id);
		int wmVersion = getVersionNumber(_id);

		return wmVersion == ownedVersion;
	}

	/**
	 * Checks whether this component has read the more recent version of the
	 * data at this working memory address, and throws and exception if not.
	 * 
	 * @param _id
	 * @throws ConsistencyException
	 *             if the id is not versioned, or if the
	 * @throws DoesNotExistOnWMException
	 *             if the _id does not exist on wm
	 * @throws SubarchitectureProcessException
	 *             if communication goes awry
	 */
	protected final void checkConsistency(String _id)
			throws ConsistencyException, DoesNotExistOnWMException,
			SubarchitectureProcessException {

		if (!isVersioned(_id)) {
			throw new ConsistencyException(new WorkingMemoryAddress(_id,
					getSubarchitectureID()), "!isVersioned(" + _id
					+ ") in subarch " + getSubarchitectureID());
		}

		if (!haveLatestVersion(_id)) {
			throw new ConsistencyException(
					new WorkingMemoryAddress(_id, getSubarchitectureID()),
					"You have attempted to overwrite an outdated working memory entry. Please reread and try again. WMA: "
							+ _id
							+ ":"
							+ getSubarchitectureID()
							+ ". Local version: "
							+ getStoredVersionNumber(_id)
							+ ". WM version: " + getVersionNumber(_id));
		}

	}

	public void setPullConnector(String _connectionID,
			PermissionsPullConnectorOut _senderAdaptor) {
		m_lockConnector = _senderAdaptor;
	}

	/**
	 * Unlock the given working memory entry.
	 * 
	 * @param _id
	 * @throws WMException
	 */
	protected void unlockEntry(String _id) throws WMException {
		unlockEntryHelper(_id, getSubarchitectureID());
	}

	/**
	 * Please ignore helper!
	 * 
	 * @param _id
	 * @param _subarch
	 * @throws WMException
	 */
	protected void unlockEntryHelper(String _id, String _subarch)
			throws WMException {

		// if
		if (!holdsLock(_id, _subarch)) {
			debug("no lock held for: " + _id + ":" + _subarch);
			return;
		}

		try {
			WorkingMemoryPermissions set = m_lockConnector
					.pull(new FrameworkQuery(getProcessIdentifier(), CASTUtils
							.unlockQueryString(_id, _subarch)));

			// allow for DOES_NOT_EXIST as this could happen if unlock on
			// deletion occurs
			// boolean succeeded = (WorkingMemoryPermissions.UNLOCKED == set
			// ||
			// WorkingMemoryPermissions.DOES_NOT_EXIST == set);

			if (set == WorkingMemoryPermissions.DOES_NOT_EXIST) {
				throw new DoesNotExistOnWMException(new WorkingMemoryAddress(
						_subarch, _id), "ID " + _id
						+ " does not exist on WM in subarch " + _subarch);
			}

			boolean succeeded = (WorkingMemoryPermissions.UNLOCKED == set);

			// we should always succeed unlocking if we hold a lock
			assert (succeeded);

			m_permissions.removePermissions(_id, _subarch);
			assert (!holdsLock(_id, _subarch));

		}
		catch (FrameworkConnectionException e) {
			throw new WMException(new WorkingMemoryAddress(_id, _subarch),
					"Error locking entry", e);
		}
	}

	/**
	 * 
	 * Try to obtain a lock on a working memory entry with the given
	 * permissions. This will block until the desired lock is obtained.
	 * 
	 * @param _id
	 *            The id of the item on working memory.
	 * @param _permissions
	 *            The permissions to obtain.
	 * @throws DoesNotExistOnWMException
	 *             If the item does not exist on wm.
	 * @throws SubarchitectureProcessException
	 *             If other errors occur.
	 */
	protected void lockEntry(String _id, WorkingMemoryPermissions _permissions)
			throws SubarchitectureProcessException {
		lockEntryHelper(_id, getSubarchitectureID(), _permissions,
				OperationMode.BLOCKING);
	}

	/**
	 * Try to obtain a lock on a working memory entry. This will return true if
	 * the item is locked, or false if not. This method does not block.
	 * 
	 * @param _id
	 *            The id of the item on working memory.
	 * @param _permissions
	 *            The permissions to obtain.
	 * @throws DoesNotExistOnWMException
	 *             If the item does not exist on wm.
	 * @throws SubarchitectureProcessException
	 *             If other errors occur.
	 */
	protected boolean tryLockEntry(String _id,
			WorkingMemoryPermissions _permissions)
			throws SubarchitectureProcessException {
		return lockEntryHelper(_id, getSubarchitectureID(), _permissions,
				OperationMode.NON_BLOCKING);
	}

	/**
	 * Map representing the permissions currently held by this component.
	 */
	private CASTComponentPermissionsMap m_permissions;

	/**
	 * Ignore helper please!
	 * 
	 * @throws WMException
	 * @throws WMException
	 */
	protected final boolean lockEntryHelper(String _id, String _subarch,
			WorkingMemoryPermissions _permissions, OperationMode _op)
			throws WMException {
		try {
			WorkingMemoryPermissions set = m_lockConnector
					.pull(new FrameworkQuery(getProcessIdentifier(), CASTUtils
							.lockQueryString(_id, _subarch, _permissions, _op)));

			if (set == WorkingMemoryPermissions.DOES_NOT_EXIST) {
				throw new DoesNotExistOnWMException(new WorkingMemoryAddress(
						_id, _subarch), "ID " + _id
						+ " does not exist on WM in subarch " + _subarch);
			}

			boolean succeeded = (_permissions == set);

			// if we blocked then we should always succeed
			if (_op == OperationMode.BLOCKING) {
				assert (succeeded);
			}

			// if we succeeded, then let's store the permissions
			if (succeeded) {
				m_permissions.setPermissions(_id, _subarch, set);
			}

			return succeeded;

		}
		catch (FrameworkConnectionException e) {
			throw new WMException(new WorkingMemoryAddress(_id, _subarch),
					"Error locking entry", e);
		}
	}

	/**
	 * Ignore helper please!
	 * 
	 * @param _id
	 * @param _subarch
	 * @return
	 * @throws WMException
	 */
	protected final WorkingMemoryPermissions getPermissionsHelper(String _id,
			String _subarch) throws WMException {
		try {
			WorkingMemoryPermissions permissions = m_lockConnector
					.pull(new FrameworkQuery(getProcessIdentifier(), CASTUtils
							.statusQueryString(_id, _subarch)));

			if (permissions == WorkingMemoryPermissions.DOES_NOT_EXIST) {
				throw new DoesNotExistOnWMException(new WorkingMemoryAddress(
						_id, _subarch), "ID " + _id
						+ " does not exist on WM in subarch " + _subarch);
			}

			return permissions;
		}
		catch (FrameworkConnectionException e) {
			throw new WMException(new WorkingMemoryAddress(_id, _subarch),
					"Error getting permissions", e);
		}
	}

	/**
	 * Checks whether a given item on working memory is currently overwritable
	 * by this component.
	 * 
	 * @param _id
	 * @return
	 * @throws WMException
	 */
	protected boolean isOverwritable(String _id) throws WMException {
		if (holdsOverwriteLock(_id)) {
			return true;
		}
		else {
			WorkingMemoryPermissions permissions = getPermissions(_id);
			return CASTUtils.overwriteAllowed(permissions);
		}
	}

	/**
	 * Checks whether a given item on working memory is currently deletable by
	 * this component.
	 * 
	 * @param _id
	 * @return
	 * @throws WMException
	 */
	protected boolean isDeletable(String _id) throws WMException {
		if (holdsDeleteLock(_id)) {
			return true;
		}
		else {
			WorkingMemoryPermissions permissions = getPermissions(_id);
			return CASTUtils.deleteAllowed(permissions);
		}
	}

	/**
	 * Checks whether a given item on working memory is currently readable by
	 * this component.
	 * 
	 * @param _id
	 * @return
	 * @throws WMException
	 */
	protected boolean isReadable(String _id) throws WMException {
		if (holdsReadLock(_id)) {
			return true;
		}
		else {
			WorkingMemoryPermissions permissions = getPermissions(_id);
			return CASTUtils.readAllowed(permissions);
		}
	}

	/**
	 * Gets the permissions currently set on the given working memory item.
	 * 
	 * @param _id
	 * @return
	 * @throws WMException
	 */
	protected WorkingMemoryPermissions getPermissions(String _id)
			throws WMException {
		return getPermissionsHelper(_id, getSubarchitectureID());
	}

	/**
	 * Checks whether this component holds any kind of lock on the given working
	 * memory item.
	 * 
	 * @param _id
	 * @return
	 */
	protected boolean holdsLock(String _id) {
		return holdsOverwriteLock(_id);
	}

	/**
	 * Checks whether this component holds any kind of lock on the given working
	 * memory item.
	 * 
	 * @param _id
	 * @param _subarch
	 * @return
	 */
	protected boolean holdsLock(String _id, String _subarch) {
		return holdsOverwriteLock(_id, _subarch);
	}

	/**
	 * Checks whether this component holds an overwrite lock on the given
	 * working memory item.
	 * 
	 * @param _id
	 * @return
	 */
	protected boolean holdsOverwriteLock(String _id) {
		// any lock is an overwrite lock
		return m_permissions.getPermissions(_id) != null;
	}

	/**
	 * Checks whether this component holds a delete lock on the given working
	 * memory item.
	 * 
	 * @param _id
	 * @return
	 */
	protected boolean holdsDeleteLock(String _id) {
		WorkingMemoryPermissions permissions = m_permissions
				.getPermissions(_id);
		return permissions != null
				&& (permissions == WorkingMemoryPermissions.LOCKED_OD || permissions == WorkingMemoryPermissions.LOCKED_ODR);
	}

	/**
	 * Checks whether this component holds a read lock on the given working
	 * memory item.
	 * 
	 * @param _id
	 * @return
	 */
	protected boolean holdsReadLock(String _id) {
		WorkingMemoryPermissions permissions = m_permissions
				.getPermissions(_id);
		return permissions != null
				&& permissions == WorkingMemoryPermissions.LOCKED_ODR;
	}

	/**
	 * Checks whether this component holds an overwrite lock on the given
	 * working memory item.
	 * 
	 * @param _id
	 * @param _subarch
	 * @return
	 */
	protected boolean holdsOverwriteLock(String _id, String _subarch) {
		// any lock is an overwrite lock
		return m_permissions.getPermissions(_id, _subarch) != null;
	}

	/**
	 * Checks whether this component holds a delete lock on the given working
	 * memory item.
	 * 
	 * @param _id
	 * @param _subarch
	 * @return
	 */
	protected boolean holdsDeleteLock(String _id, String _subarch) {
		WorkingMemoryPermissions permissions = m_permissions.getPermissions(
				_id, _subarch);
		return permissions != null
				&& (permissions == WorkingMemoryPermissions.LOCKED_OD || permissions == WorkingMemoryPermissions.LOCKED_ODR);
	}

	/**
	 * Checks whether this component holds a read lock on the given working
	 * memory item.
	 * 
	 * @param _id
	 * @param _subarch
	 * @return
	 */
	protected boolean holdsReadLock(String _id, String _subarch) {
		WorkingMemoryPermissions permissions = m_permissions.getPermissions(
				_id, _subarch);
		return permissions != null
				&& permissions == WorkingMemoryPermissions.LOCKED_ODR;
	}

	protected boolean needsConsistencyCheck(String _id) {
		return m_permissions.needsConsistencyCheck(_id);
	}

	protected boolean needsConsistencyCheck(String _id, String _subarch) {
		return m_permissions.needsConsistencyCheck(_id, _subarch);
	}

	protected void consistencyChecked(String _id) {
		m_permissions.consistencyChecked(_id);
	}

	protected void consistencyChecked(String _id, String _subarch) {
		m_permissions.consistencyChecked(_id, _subarch);
	}

}
